### ISSUE TITLE
Add more shape recognition options to the shape recognizer.

### ISSUE THREAD

**Issue by K4LCIFER** [xournalpp/xournalpp#3665]
**Is your feature request related to a problem? Please describe.**
Currently it seems that the shape recognizer only recognizes rectangles, squares, and perfect circles.

**Describe the solution you'd like**
It would be more beneficial if the shape recognizer were able to recognize *all* basic shapes, i.e.
- triangles (right, isoceles, etc)
- eliipses

**Additional Context**
If possible, it would be nice to have a more general shape recognition tool that takes the form of predictive smoothing. For exampe, If I draw some function (parabaloid, exponential, sinusoid, etc), the shape recognizer should be able to smooth out my lines so that the functions follow the shape that they would mathematically follow.

I don't think there is any need for polynomial function recognition beyond a cubic function, but it would be good to support all of the basic functions:
- polynomial up to 3rd order
- sinusoid
- exponential/logarithmic
- etc

**Comment by bhennion** [comment]
See also #2275 and the discussion therein.

**Comment by lluisalemanypuig** [comment]
I agree: it would be nice if triangles (and other shapes) were recognized by the shape recognizer. So I thought maybe I could start with the triangles, and eventually did something that works. Here's a [video](https://youtu.be/7UuRd2_55MQ) that shows the result.

However, I'm not sure I understood the task properly. The rectangles, for example, are created with perfect 90º-angles. Triangles can be idealized (with, say, 60º angles, or detect when a triangle has a specific configuration of angles) but then the recognizer would recognize only so many triangles. The fix I have (for which I do not currently have a PR) creates triangles by ensuring that the three strokes make up a triangle, and uses the segments to create a single stroke, whichever resulting triangle we get. Are there any triangles that should _not_ be recognized?

Also, I think I can generalize what I've done to polygonal strokes. But I'm not sure this would be interesting since the only thing that the recognizer would do is straighten up the strokes.
